File: gruver/tune.php
<?php

function detectCirclesPHP($im, $w, $h, $min_rad_px, $max_rad_px)
{
    $found = [];
    $step = 3;

    // We want to debug a bit more closely, so we will generate an output image
    $dbg_im = imagecreatetruecolor($w, $h);
    imagecopy($dbg_im, $im, 0, 0, 0, 0, $w, $h);
    $red = imagecolorallocate($dbg_im, 255, 0, 0);

    // Calculate radii to test (between min and max)
    $radii = [];
    for ($r = $min_rad_px; $r <= $max_rad_px; $r++) {
        $radii[] = $r;
    }
    if (empty($radii))
        $radii = [5, 6, 7];

    foreach ($radii as $r) {
        for ($y = $r + 5; $y < $h - $r - 5; $y += $step) {
            for ($x = $r + 5; $x < $w - $r - 5; $x += $step) {
                $center = imagecolorat($im, $x, $y) & 0xFF; // Already grayscale

                $num_samples = 16;
                $hits = 0;
                $total_contrast = 0;
                // INCREASED THRESHOLD drastically
                $thresh = 26;

                $quadrants = [0, 0, 0, 0];
                for ($i = 0; $i < $num_samples; $i++) {
                    $a = ($i / $num_samples) * 2 * M_PI;
                    $sx = (int) ($x + cos($a) * $r);
                    $sy = (int) ($y + sin($a) * $r);

                    $v = imagecolorat($im, $sx, $sy) & 0xFF;
                    $diff = abs($v - $center);

                    // We also want the rim to be consistently DARKER or LIGHTER?
                    // Usually the rim of a pit is lighter (if light hits) or darker (shadow)
                    if ($diff > $thresh) {
                        $hits++;
                        $total_contrast += $diff;
                        $quad = (int) floor(($a / (2 * M_PI)) * 4);
                        if ($quad > 3)
                            $quad = 3;
                        $quadrants[$quad]++;
                    }
                }

                $active_quads = 0;
                foreach ($quadrants as $q) {
                    if ($q >= 1)
                        $active_quads++;
                }

                // STRENGTHENED REQUIREMENT
                if ($hits >= 12 && $active_quads == 4) {
                    $prob = 0.5 + ($hits / 32) + (min($total_contrast, 300) / 1000);
                    $found[] = ['x' => $x, 'y' => $y, 'cat' => 'kullmiler', 'prob' => min($prob, 0.98)];
                    imagearc($dbg_im, $x, $y, $r * 2, $r * 2, 0, 360, $red);
                    $x += 4;
                }
            }
        }
    }
    imagepng($dbg_im, "data/funn/tuned_output.png");
    return $found;
}

$img_data = file_get_contents('https://voxcuriosa.no/gruver/data/funn/debug_tile.png');
$src = imagecreatefromstring($img_data);
echo "Hits: " . count(detectCirclesPHP($src, 512, 512, 5, 10)) . "\n";
?>

File: gruver/tune.py
import cv2
import numpy as np
import urllib.request
import os

def detect_circles_php_style(img_path, min_rad, max_rad):
    img = cv2.imread(img_path)
    gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
    # The PHP script applies -15 contrast which maps roughly to 0.85 alpha in OpenCV
    # but we'll just use the raw grayscale for now as it's close enough.
    h, w = gray.shape

    out_img = img.copy()
    found = 0
    
    radii = list(range(min_rad, max_rad + 1))
    step = 3
    
    thresh = 22 # Testing threshold
    
    for r in radii:
        for y in range(r + 5, h - r - 5, step):
            for x in range(r + 5, w - r - 5, step):
                center = int(gray[y, x])
                
                num_samples = 16
                hits = 0
                total_contrast = 0
                quadrants = [0, 0, 0, 0]
                
                for i in range(num_samples):
                    a = (i / num_samples) * 2 * np.pi
                    sx = int(x + np.cos(a) * r)
                    sy = int(y + np.sin(a) * r)
                    
                    v = int(gray[sy, sx])
                    diff = abs(v - center)
                    
                    if diff > thresh:
                        hits += 1
                        total_contrast += diff
                        quad = int((a / (2 * np.pi)) * 4)
                        if quad > 3: quad = 3
                        quadrants[quad] += 1
                        
                active_quads = sum(1 for q in quadrants if q >= 1)
                
                if hits >= 12 and active_quads == 4:
                    cv2.circle(out_img, (x, y), r, (0, 0, 255), 1)
                    found += 1
                    
    print(f"Hits with thresh={thresh}, hits>=12: {found}")
    cv2.imwrite("data/funn/tuned_output.png", out_img)

if not os.path.exists("data/funn/debug_tile.png"):
    urllib.request.urlretrieve('https://voxcuriosa.no/gruver/data/funn/debug_tile.png', 'data/funn/debug_tile.png')
    
detect_circles_php_style("data/funn/debug_tile.png", 5, 10)
